### ISSUE TITLE
Set-OvUplinkSet : Support of NetworkSets when modifying an uplinkset in a LIG

### ISSUE THREAD

**Issue by odesaphy** [HewlettPackard/POSH-HPOneView#512]
# Summary of the new feature/enhancement

Since OneView 5.00 we can add a NetworkSet to an UplinkSet when creating/modifying a LIG 

# Proposed technical implementation details (optional)

New parameter -AddNetworkSet
```powershell
Get-OvLogicalInterconnectGroup -Name MY_VC_LIG | Set-OVUplinkSet -UplinkSetName MY_UPLINKSET -AddNetworkSet MY_NETWORKSET
```
New parameter -RemoveNetworkSet
```powershell
Get-OvLogicalInterconnectGroup -Name MY_VC_LIG | Set-OVUplinkSet -UplinkSetName MY_UPLINKSET -RemoveNetworkSet MY_NETWORKSET
```

**Comment by ChrisLynchHPE** [comment]
This will be added to the 5.00 and newer libraries.

**Comment by odesaphy** [comment]
I've found in the Set-OVUplinkSet function there are several references to variables not used : 

For example on HPEOneView.540 5.40.2559.2085 - HPEOneView.540.psm1
Line 94759 : $NetworkSetsUri
Line 94946 : $_UpdatedNetSet,

**Comment by ChrisLynchHPE** [comment]
This is now addressed in the following releases:

* [5.50.2680.1687](https://www.powershellgallery.com/packages/HPEOneView.550/5.50.2680.1687) 
* [5.40.2681.1725](https://www.powershellgallery.com/packages/HPEOneView.540/5.40.2681.1725)
* [5.30.2682.3811](https://www.powershellgallery.com/packages/HPEOneView.530/5.30.2682.3811)
* [5.20.2684.2161](https://www.powershellgallery.com/packages/HPOneView.520/5.20.2684.2161)
* [5.00.2685.1865](https://www.powershellgallery.com/packages/HPOneView.500/5.0.2685.1865)

Please validate prior to closing this issue.

**Comment by odesaphy** [comment]
Only the NetworkSet uri is added to the UplinkSet.networkSetUris

UplinkSet.networkUris is not updated with the uris of the networks in the networkset 

and so we have this error code : CRM_NETWORKSET_NETWORKS_IS_NOT_FOUND_IN_UPLINKSET_IN_LIG

**Comment by ChrisLynchHPE** [comment]
Hmmm....  The API behavior was changed in newer releases.  Which library release are you using?

**Comment by odesaphy** [comment]
I've used latest 5.50 library connected to 6.0 dcs

From the rest API reference LIG
networkUris : 
A set of network set URIs assigned to the uplink set. The list can be empty but not null. NOTE: When adding a network set, also add its constituent networks to networkUris. When removing a network set, also remove its constituent networks from networkUris, unless they are to remain configured in the uplink set.

**Comment by odesaphy** [comment]
I've tested with HPEOneView 6.10.2778.2020

And it's not working.

The error is in the ForEach loop adding network uris (from the networkset) in $_UpdatedNetworkUris

In the following loop $_net is a uri and so adding $_net.uri to the array doesn't work.

```
ForEach ($_net in $_NetSetToAdd.networkUris)
                    {
                        if (-not $_UpdatedNetworkUris.Contains($_net.uri))
                        {
                            [void]$_UpdatedNetworkUris.Add($_net.uri)
                        }
                    }
```

Fix : 

```
ForEach ($_netUri in $_NetSetToAdd.networkUris)
                    {
                        if (-not $_UpdatedNetworkUris.Contains($_netUri ))
                        {
                            [void]$_UpdatedNetworkUris.Add($_netUri)
                        }
                    }
```

**Comment by odesaphy** [comment]
I've tested successfully the AddNetworkSet usecase with HPEOneView.630 library (6.30.2913.2047) 

The RemoveNetworkSet usecase is not working
`Get-OvLogicalInterconnectGroup -Name MY_VC_LIG | Set-OVUplinkSet -UplinkSetName MY_UPLINKSET -RemoveNetworkSet MY_NETWORKSET`
Only NetworkSet is removed but the networks in the network set are not removed from the UplinkSet

**Comment by ChrisLynchHPE** [comment]
This is fixed in [Release 6.30.3188.1900](https://github.com/HewlettPackard/POSH-HPEOneView/releases/tag/v6.30.3188.1900) and in [PowerShell Gallery](https://www.powershellgallery.com/packages/HPEOneView.630/6.30.3188.1900).

Please verify before this issue is closed.

**Comment by odesaphy** [comment]
In the "ForEach ($_netUriToRemove in $_NetSetToRemove.networkUris)" loop there is a lookup via index for net to netset association

If the result count is > 0 then the network is not removed from the uplinkset (not removed from $_UpdatedNetworkUris)
but as we looping through the networks of a networkset the result count of NETWORKSET_TO_NETWORK will be always >= 1

So if result count is > 1 (network is also part of other(s) networkset(s) we do not remove it from the uplinkset

Proposition :

Change  
```
if ($_IndexResults.count -gt 0)
{
    "[{0}] Network is associated with {1} exsiting network set(s).  Will not remove." -f $MyInvocation.InvocationName.ToString().ToUpper(), $_IndexResults.count | Write-Verbose
}
else
{
    "[{0}] Removing network from uplink set." -f $MyInvocation.InvocationName.ToString().ToUpper() | Write-Verbose
    [void]$_UpdatedNetworkUris.Remove($_netUriToRemove)
}
```

To
```
if ($_IndexResults.count -gt 1)
{
    "[{0}] Network is associated with {1} exsiting network set(s).  Will not remove." -f $MyInvocation.InvocationName.ToString().ToUpper(), $_IndexResults.count | Write-Verbose
}
else
{
    "[{0}] Removing network from uplink set." -f $MyInvocation.InvocationName.ToString().ToUpper() | Write-Verbose
    [void]$_UpdatedNetworkUris.Remove($_netUriToRemove)
}
```

**Comment by ChrisLynchHPE** [comment]
What you are proposing isn't correct.  You can assign multiple network set resources to a single uplink set, where a network can be assigned to one or more of those network sets.  The logic here is to ensure that the networks within the network set being removed aren't removed by accident when there are multiple network sets assigned.

After looking at this further, the evaluation logic here isn't quite right either, as it is only looking to see if the network is a member of more than 1 network set, but then doesn't evaluate if the network set is also assigned to the uplink set.  Once that is added, then the logic flow here should then work.

**Comment by ChrisLynchHPE** [comment]
This is fixed in Release 6.60.3227.2799, published in the [PowerShell Gallery](https://www.powershellgallery.com/packages/HPEOneView.660/6.60.3227.2799).

Closing due to no further activity.  If you wish, feel free to re-open to discuss further.
